Lib/fontTools/varLib/hvar.py

```diff
@@ -0,0 +1,167 @@
+from fontTools.misc.roundTools import noRound
+from fontTools.ttLib import TTFont, newTable
+from fontTools.ttLib.tables import otTables as ot
+from fontTools.ttLib.tables.otBase import OTTableWriter
+from fontTools.varLib import HVAR_FIELDS, VVAR_FIELDS
+from fontTools.varLib import builder, varStore
+from fontTools.misc.fixedTools import fixedToFloat as fi2fl
+from fontTools.misc.cliTools import makeOutputFileName
+from itertools import product
+import logging
+
+log = logging.getLogger("fontTools.varLib.avar")
+
+
+def add_VHVAR(font, tableFields):
+    tableTag = tableFields.tableTag
+    if tableTag in font:
+        log.warning(tableTag + " table already present, overwriting.")
+        del font[tableTag]
+    log.info("Generating " + tableTag)
+    VHVAR = newTable(tableTag)
+    tableClass = getattr(ot, tableTag)
+    vhvar = VHVAR.table = tableClass()
+    vhvar.Version = 0x00010000
+    font[tableTag] = VHVAR
+
+    glyphOrder = font.getGlyphOrder()
+    vhmtx = font[tableFields.metricsTag]
+    advances = {g: vhmtx.metrics[g][0] for g in glyphOrder}
+    phantomIndex = tableFields.phantomIndex
+
+    axisTags = [a.axisTag for a in font["fvar"].axes]
+    storeBuilder = varStore.OnlineVarStoreBuilder(axisTags)
+
+    # There's two ways we can go from here:
+    # 1. For each glyph, at each master peak, compute the value of the
+    #    advance width at that peak.  Then pass these all to a VariationModel
+    #    builder to compute back the deltas.
+    # 2. For each master peak, pull out the deltas of the advance width directly,
+    #    and feed these to the VarStoreBuilder, forgoing the remoding step.
+    # We'll go with the second option, as it's simpler, faster, and more direct.
+    gvar = font["gvar"]
+    advMapping = {}
+    allDeltas = {}
+    for glyphName in glyphOrder:
+        supports = []
+        deltas = []
+        variations = gvar.variations[glyphName]
+
+        for tv in variations:
+            supports.append(tv.axes)
+            phantoms = tv.coordinates[-4:]
+            phantoms = phantoms[phantomIndex * 2 : phantomIndex * 2 + 2]
+            assert len(phantoms) == 2
+            phantoms[0] = phantoms[0][phantomIndex] if phantoms[0] is not None else 0
+            phantoms[1] = phantoms[1][phantomIndex] if phantoms[1] is not None else 0
+            deltas.append(phantoms[1] - phantoms[0])
+
+        allDeltas[glyphName] = deltas[1:]
+        storeBuilder.setSupports(supports)
+        advMapping[glyphName] = storeBuilder.storeDeltas(deltas, round=noRound)
+
+    singleModel = all(advMapping[g] >> 16 == 0 for g in glyphOrder)
```

Comment:
shouldn't this check that all the supports are the same (like we do in the varLib _get_advance_metrics)?

Reply:
That's one way. My reasoning here was that if the OnlineVarStoreBuilder put them all in the same group (ie. all major's match), then we can do this. I'll try to revert to other logic.